db/clientlist_virtual_reader.cc

```diff
@@ -0,0 +1,237 @@
+    }).then([] (distributed<cql_transport::cql_server>& distributed_cql_server) {
+        return get_all_clients_data(distributed_cql_server);
+    }).then([this] (auto&& client_data_vec) {
+        return filter_and_sort_by_port(std::move(client_data_vec)).then([this] (auto&& sorted_client_data_vec) {
+            return do_with(
+                    std::vector<mutation>{},
+                    [this, scdv=std::move(sorted_client_data_vec)] (auto&& mutations_vec) mutable {
+                mutations_vec.reserve(scdv.size());
+
+                std::for_each(scdv.begin(), scdv.end(), [this, &mutations_vec] (auto&& client) {
+                    mutations_vec.emplace_back(db::system_keyspace::make_client_mutation(std::move(client)));
+                });
+                if (!mutations_vec.empty()) {
+                    _mutation_fragmentator_opt = flat_mutation_reader_from_mutations(std::move(mutations_vec));
+                }
+                return make_ready_future<>();
+            });
+        });
+    });
+}
+
+clientlist_mutation_reader::clientlist_mutation_reader(
+        schema_ptr schema,
+        const dht::partition_range& prange,
+        const query::partition_slice& slice)
+    : impl(schema)
+    , _schema(std::move(schema))
+    , _prange(&prange)
+    , _slice(slice) {}
+
+future<> clientlist_mutation_reader::fill_buffer(db::timeout_clock::time_point timeout) {
+    // PKs returned from different shards must be distinct, otherwise combining reader
+    // hangs up. Now: one client can maintain concurrent connections with many different
+    // shards but his IP belongs to partition range of a particular (possibly yet another)
+    // shard. That's why every shard must traverse every connection and return only the
+    // clients with IPs that fall within *this shard's* partition range.
+
+    return with_timeout(timeout, maybe_setup_mutation_fragmentator()).then([this, timeout] {
+        if (!_mutation_fragmentator_opt) {
+            // Setup of `_mutation_fragmentator_opt' failed => there are no connections
+            _end_of_stream = true;
+            return make_ready_future<>();
+        }
+        return do_until([this, timeout] { return is_end_of_stream() || is_buffer_full(); }, [this, timeout] {
+            return _mutation_fragmentator_opt->consume_pausable([this] (mutation_fragment mf) {
+                push_mutation_fragment(std::move(mf));
+                return stop_iteration(is_buffer_full());
+            }, timeout).then([this] {
+                if (_mutation_fragmentator_opt->is_end_of_stream() && _mutation_fragmentator_opt->is_buffer_empty()) {
+                    _end_of_stream = true;
+                    _mutation_fragmentator_opt = std::nullopt;
+                }
+            });
+        });
+    });
+}
+
+void clientlist_mutation_reader::next_partition() {
+    // This method should never be needed.
+    api::unimplemented();
```

Comment:
If the comment is left like this, then I agree. Simply throwing `std::bad_function_call()` and deleting this comment would be equally bad to throwing and leaving this comment.

What would be infinitely times better is to explain **why** this method should never be needed.
For some reason `bad_function_call` doesn't take a string argument, so the explanation should be written in the comment.

Reply:
For me no explanation is needed (maybe because I saw many readers). Readers, that don't support fast-forwarding (neither of them) don't allow these methods to be called. IMO this is implicit, we have many such readers in the codebase. This is also part of the `flat_mutation_reader` interface contract, although it might not be documented.

Reply:
Why doesn't this reader support fast-forwarding?

Reply:
Fast-forwarding is only used by the cache, so only readers that are used by cache as an underlying reader need to support it. I think it doesn't make sense to cache virtual tables, but I'm not sure if we actually have an optimization for that. @tgrabiec?

Reply:
And that deserves a comment.

Reply:
Reads for tables with virtual readers don't go through cache currently.